### PR TITLE
Expose more variables in jbuild files

### PR DESCRIPTION
**Pull request by diml**
These are needed when linking executables as objects/shared objects (#23): `ext_obj`, `ext_asm`,  `ext_lib`, `ext_dll` and `ext_exe`, `bytecomp_c_libraries` and  `native_c_libraries`.

I also moved `Ocamlc_config` to its own library

### CODE REVIEW

File: src/ocamlc-config/ocamlc_config.mli [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+type t
+val sexp_of_t : t -> Usexp.t
+
+exception E of string
+
+(** Create a [t] value from the output of [ocamlc -config]. Return
+    [Error msg] in case of error. *)
+val of_lines : string list -> t
+
+(** All the uninterpret variables in the output [ocamlc -configurator] *)
```
Comment (rgrinberg) [inline src/ocamlc-config/ocamlc_config.mli]: You mean `-config` here.

**Comment by rgrinberg** [comment]
Can we also drop the dune dependency in configurator then? Now we can depend on just ocamlc-config.

File: src/ocamlc-config/ocamlc_config.mli (hunk 1 above)
Comment (diml) [reply src/ocamlc-config/ocamlc_config.mli]: indeed

**Comment by diml** [comment]
Not yet, it still uses `Io`

**Comment by dra27** [comment]
Could we also expose `ccomp_type` while we're here?

**Comment by diml** [comment]
@dra27 how can we use it at the moment? Is it to do `(run ... -ccomp-type ${ccomp_type})`?

**Comment by diml** [comment]
BTW, one thing I was wondering is to just expose `${ocaml-config:XXX}`. The only problem with that is that the set of variables is not stable

**Comment by rgrinberg** [comment]
Why is `${ocamlc-config:XXX}` not stable? We can still post-process it however we want to keep it constant/consistent across OCaml versions.

**Comment by dra27** [comment]
I'd been musing on having all the variables too! I agree with @rgrinberg - it's relatively easy to emulate missing ones in the versions of OCaml we already support (we already have to do it for the c compiler)

**Comment by dra27** [comment]
Regarding, `ccomp_type`, I was thinking it might be usable to select between different files for C flags, although I can't remember if we can put variables into file names like that? I'm impaired by a 2yo at the moment for trying to look at the code I had in mind!

**Comment by diml** [comment]
I suppose, yh. I'll update this PR. I think we can keep the `ext_...` vars as it though.

> Regarding, ccomp_type, I was thinking it might be usable to select between different files for C flags, although I can't remember if we can put variables into file names like that? I'm impaired by a 2yo at the moment for trying to look at the code I had in mind!

Ah, indeed. Yh that's supported

**Comment by diml** [comment]
Ok, I did more work to produce a stable set of key-value pairs. Now the raw output is completely hidden. All variables are now accessibly through `${ocaml-config:<var>}`. I also renamed the library and module `Ocaml_config`.

I think we could remove many of the fields from `Context.t` and only leave `ocaml_config`, to avoid duplication. But I've spend enought time on this so I'll leave this for future work.

**Comment by rgrinberg** [comment]
OK, this is good. What's a bit of a shame is that we're not representing some of those variables as `Path.t` in the api. Going back to untyped paths is a bit hard after getting used to how it's done in dune.

Maybe we should consider adding a subset of Path.t to stdune? We'd need to split out all the dune specific stuff from the Path module first however.

**Comment by diml** [comment]
Indeed, we should probably do something like this.

**Pull request merged**